libs/sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java

```diff
@@ -2290,17 +2293,24 @@ public Node visitEscapedCharsStringLiteral(SqlBaseParser.EscapedCharsStringLiter
     @Override
     public Node visitIntegerLiteral(SqlBaseParser.IntegerLiteralContext context) {
         String text = context.getText().replace("_", "");
-        long value = Long.parseLong(text);
-        if (value < Integer.MAX_VALUE + 1L) {
-            return new IntegerLiteral((int) value);
+        BigInteger bigInteger = new BigInteger(text);
+        int bitLength = bigInteger.bitLength();
+        if (bitLength <= 31) {
+            return new IntegerLiteral(bigInteger.intValueExact());
+        } else if (bitLength <= 63) {
+            return new LongLiteral(bigInteger.longValueExact());
         }
-        return new LongLiteral(value);
+        return new NumericLiteral(new BigDecimal(bigInteger, 0));
     }
 
     @Override
     public Node visitDecimalLiteral(SqlBaseParser.DecimalLiteralContext context) {
         String text = context.getText().replace("_", "");
-        return new DoubleLiteral(text);
+        BigDecimal bigDecimal = new BigDecimal(text);
```

Comment:
Note for the future: If this causes problems, we could split up the grammar in the parser to distinguish this
Instead of:

```
DECIMAL_VALUE
    : INTEGER_VALUE '.' INTEGER_VALUE?
    | '.' DIGIT+
    | INTEGER_VALUE ('.' DIGIT*)? EXPONENT
    | '.' DIGIT+ EXPONENT
    ;
```

We could have:


```
DECIMAL_VALUE
    : DECIMAL_INTEGER '.' DECIMAL_INTEGER?
    | '.' DECIMAL_INTEGER
    ;

DOUBLE_VALUE
    : DIGIT+ ('.' DIGIT*)? EXPONENT
    | '.' DIGIT+ EXPONENT
    ;
```